### PR TITLE
FLUME-3027 Kafka Channel should clear offsets map after commit

### PR DESCRIPTION
**Pull request by jholoman**
Added a call to clear the offsets map after a commit so as to avoid repeatedly committing already-committed offsets. Also updated various debug and trace log messages/calls to help with troubleshooting.

### CODE REVIEW

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java [hunk 1]
```diff
@@ -564,14 +564,22 @@ protected void doCommit() throws InterruptedException {
                   ex);
         }
       } else {
+        //event taken ensures that we have collected events in this transaction
```
Comment (simonati) [inline flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java]: nit: missing space after //

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java [hunk 2]
```diff
@@ -564,14 +564,22 @@ protected void doCommit() throws InterruptedException {
...
+        // before committing
         if (consumerAndRecords.get().failedEvents.isEmpty() && eventTaken) {
```
Comment (simonati) [inline flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java:566]: Naive question: could you please tell why the else path is not needed. Condition to the if statement suggests that we silently ignore some failure scenario.

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java [hunk 3]
```diff
@@ -677,34 +685,45 @@ private Event deserializeValue(byte[] value, boolean parseAsFlumeEvent) throws I
...
+    String getCommittedOffsetsString() {
       StringBuilder sb = new StringBuilder();
+      sb.append("Channel ").append(this.uuid)
+          .append(" committed: ");
       for (TopicPartition tp : this.consumer.assignment()) {
         try {
-          sb.append("Committed: [").append(tp).append(",")
+          sb.append("[").append(tp).append(",")
               .append(this.consumer.committed(tp).offset())
-              .append(",").append(this.consumer.committed(tp).metadata()).append("]");
```
Comment (simonati) [inline flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java:697]: is consumer.committed(tp).metadata() not useful anymore?

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java [hunk 4]
```diff
@@ -677,34 +685,45 @@ private Event deserializeValue(byte[] value, boolean parseAsFlumeEvent) throws I
...
+      for (TopicPartition tp : offsets.keySet()) {
+        sb.append("p").append(tp.partition()).append("-")
+            .append(offsets.get(tp).offset()).append(" ");
+      }
+      return sb.toString();
+    }
 
-    public void printCurrentAssignment() {
+    // This prints the current committed offsets when debug is enabled
+    String getCommittedOffsetsString() {
```
Comment (simonati) [inline flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java]: I think this should be private

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java [hunk 5]
```diff
@@ -677,34 +685,45 @@ private Event deserializeValue(byte[] value, boolean parseAsFlumeEvent) throws I
...
-      logger.trace("polling");
+      logger.debug("returned {} records from last poll channel-{}", records.count(), channelUUID);
     }
 
     void commitOffsets() {
       this.consumer.commitSync(offsets);
     }
 
-    // This will reset the latest assigned partitions to the last committed offsets;
+    String getOffsetMapString() {
```
Comment (simonati) [inline flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java]: I think this should be private

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java [hunk 6]
```diff
@@ -564,14 +564,22 @@ protected void doCommit() throws InterruptedException {
...
+        //event taken ensures that we have collected events in this transaction
+        // before committing
         if (consumerAndRecords.get().failedEvents.isEmpty() && eventTaken) {
+          logger.trace("about to commit batch.");
           long startTime = System.nanoTime();
           consumerAndRecords.get().commitOffsets();
           long endTime = System.nanoTime();
           counter.addToKafkaCommitTimer((endTime - startTime) / (1000 * 1000));
-          consumerAndRecords.get().printCurrentAssignment();
+          counter.addToEventTakeSuccessCount(Long.valueOf(events.get().size()));
```
Comment (simonati) [inline flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java]: nit: could you please remove Long.valueOf? it is not needed (here and in the doRollback function)

**Review by simonati — CHANGES_REQUESTED** [review]
1) Log statement changes make sense to me. 

2) I would like to suggest calling the clear from the commitOffsets function within the ConsumerAndRecords class. I think this has to be addressed.

3) I would like to understand why we don't have an ``else`` path in doCommit to cover the ``failure`` scenario.

4) ConsumerAndRecords private inner class has already grown to the size which might worth the consideration of transferring it to a separate class. (As it has been already touched by this change this might be a good time for some cleanup. It won't hurt my feelings if not this time.)

5) The ``clear`` functionality might need a test. I don't know whether kafka provide an easy way to test this seemingly internal behaviour. If I got it correctly it is about giving a halfway filled topicpartition<->offset mappings as argument when calling clear after commit. This essentially translates the state-of-the-world-offset map to a not-yet-committed-offsets map. What do you think?

Haven't run the change yet.

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java (hunk 3 above)
Comment (jholoman) [reply flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java:697]: I haven't found a good use for it, and it's just polluting the logs.

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java (hunk 5 above)
Comment (jholoman) [reply flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java]: Agree

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java (hunk 6 above)
Comment (jholoman) [reply flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java]: Sure

**Comment by jholoman** [comment]
Thanks @satiatcloudera. 

2) Totally agree. I had a version with this and somehow it got reverted.
3) process() from a sink calls channel.take() until the batch size. If there is no event we break out of the loop and then commit the transaction. See https://github.com/apache/flume/blob/trunk/flume-ng-sinks/flume-hdfs-sink/src/main/java/org/apache/flume/sink/hdfs/HDFSEventSink.java#L362. 

The logic here is a little complicated, and it's always worth another look. We could probably catch the Exceptions that the Kafka Consumer could throw during commitSync CommitFailedException and Wakeup exception. Is it ok if we address this logic in another jira? 

4) I agree, but I'd like to separate that change. 

5) We are just clearing our internal tracking Map. This essentially prevents committing previously committed offsets. In a topic with a large number of partitions, this can be problematic. 

All of this is kind of hard to follow. I think the unit-tests are kind of tricky to understand, and the logic could be clearer. 

The intent of this patch is to address a situation where the Channel can get totally stuck. Happy to take another pass at refactoring the channel.

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java (hunk 6 above)
Comment (jholoman) [inline flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java]: This needs to be moved back out of the if statement.

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java (hunk 2 above)
Comment (jholoman) [reply flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java:566]: See comment

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java [hunk 7]
```diff
@@ -497,21 +495,18 @@ protected Event doTake() throws InterruptedException {
             e = deserializeValue(record.value(), parseAsFlumeEvent);
             TopicPartition tp = new TopicPartition(record.topic(), record.partition());
             OffsetAndMetadata oam = new OffsetAndMetadata(record.offset() + 1, batchUUID);
+            //we should refactor this into a method in the subclass
             consumerAndRecords.get().offsets.put(tp, oam);
 
-            if (logger.isTraceEnabled()) {
```
Comment (simonati) [inline flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java:502]: Have you measuered the performance impact of this change? Based on general guidance this should be the other way around except perf measurement shows otherwise. I would like to ask you moving back the log level checking condition if you do relatively heavy computation for calculating the argument list for logging function calls on critical paths (doTake is considered as critical path)

Please see: http://logging.apache.org/log4j/1.2/manual.html#performance

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java [hunk 8]
```diff
@@ -154,8 +155,7 @@ public void stop() {
     producer.close();
     counter.stop();
     super.stop();
-    logger.info("Kafka channel {} stopped. Metrics: {}", getName(),
```
Comment (simonati) [inline flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java:157]: I would like to ask you to revert this line. I consider shutdown metrics to be quite useful.

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java [hunk 9]
```diff
@@ -497,21 +495,18 @@ protected Event doTake() throws InterruptedException {
             e = deserializeValue(record.value(), parseAsFlumeEvent);
             TopicPartition tp = new TopicPartition(record.topic(), record.partition());
             OffsetAndMetadata oam = new OffsetAndMetadata(record.offset() + 1, batchUUID);
+            //we should refactor this into a method in the subclass
```
Comment (simonati) [inline flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java]: nit: missing whitespace after //

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java [hunk 10]
```diff
@@ -482,11 +483,8 @@ protected Event doTake() throws InterruptedException {
       if (!consumerAndRecords.get().failedEvents.isEmpty()) {
         e = consumerAndRecords.get().failedEvents.removeFirst();
       } else {
-
-        if (logger.isDebugEnabled()) {
```
Comment (simonati) [inline flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java:486]: Please see my comment on logging performance.

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java [hunk 11]
```diff
@@ -497,21 +495,18 @@ protected Event doTake() throws InterruptedException {
...
             if (record.key() != null) {
               e.getHeaders().put(KEY_HEADER, record.key());
             }
 
-            if (logger.isDebugEnabled()) {
-              logger.debug("Processed output from partition {} offset {}",
-                           record.partition(), record.offset());
-            }
+            logger.debug("Processed output from partition {} offset {}",
+                record.partition(), record.offset());
```
Comment (simonati) [inline flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java]: This looks good as partition() and offset() only returns a value and performs no computation.

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java [hunk 12]
```diff
@@ -564,15 +560,22 @@ protected void doCommit() throws InterruptedException {
...
+        // event taken ensures that we have collected events in this transaction
+        // before committing
         if (consumerAndRecords.get().failedEvents.isEmpty() && eventTaken) {
+          logger.trace("About to commit batch.");
           long startTime = System.nanoTime();
           consumerAndRecords.get().commitOffsets();
           long endTime = System.nanoTime();
           counter.addToKafkaCommitTimer((endTime - startTime) / (1000 * 1000));
-          consumerAndRecords.get().printCurrentAssignment();
+          logger.debug("{}", consumerAndRecords.get().getCommittedOffsetsString());
```
Comment (simonati) [inline flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java]: Please see my comment on logging performance.

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java [hunk 13]
```diff
@@ -373,6 +372,7 @@ private void migrateOffsets() {
   }
 
   private void decommissionConsumerAndRecords(ConsumerAndRecords c) {
+    c.consumer.wakeup();
```
Comment (simonati) [inline flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java:375]: What is the intention here?
From the javadoc of KafkaConsumer.close() it seems that wakeup won't interrupt the stop. So I would assume something else should have been interrupted by this statement. In that case wouldn't be the desired order: wakeup first then close?

> org.apache.kafka.clients.consumer.KafkaConsumer
> public void close()
> Close the consumer, waiting indefinitely for any needed cleanup. If auto-commit is enabled, this will commit the current offsets. Note that wakeup() cannot be use to interrupt close.

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java (hunk 3 above)
Comment (simonati) [reply flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java:697]: Thanks for the explanation. Fine by me.

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java [hunk 14]
```diff
@@ -564,15 +560,22 @@ protected void doCommit() throws InterruptedException {
...
+          logger.trace("About to commit batch.");
           long startTime = System.nanoTime();
           consumerAndRecords.get().commitOffsets();
           long endTime = System.nanoTime();
           counter.addToKafkaCommitTimer((endTime - startTime) / (1000 * 1000));
-          consumerAndRecords.get().printCurrentAssignment();
+          logger.debug("{}", consumerAndRecords.get().getCommittedOffsetsString());
+        }
+
+        int takes = events.get().size();
```
Comment (simonati) [inline flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java:577]: Have you considered moving this into the 
> if (consumerAndRecords.get().failedEvents.isEmpty() && eventTaken) {

block?

**Review by simonati — CHANGES_REQUESTED** [review]
Thank you for the update! The already discussed items look good these are only findings related to additional changes added after prev review.

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java (hunk 7 above)
Comment (jholoman) [reply flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java:502]: My understanding is that using the {} means they aren't evaluated if the logging mode doesn't match. Am I wrong about this?

http://www.slf4j.org/faq.html#logging_performance

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java (hunk 8 above)
Comment (jholoman) [reply flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java:157]: They are already printed in the Flume Agent log when the channel closes:

org.apache.flume.instrumentation.MonitoredCounterGroup: Shutdown Metric for type: CHANNEL, name: kafka-channel-1. channel.start.time == 1480359313437

So this was redundant.

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java (hunk 13 above)
Comment (jholoman) [reply flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java:375]: The intention is to break out of the blocking thread in poll(), before closing the consumer.

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java (hunk 14 above)
Comment (jholoman) [reply flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java:577]: Maybe? I don't have a strong opinion on it and would prefer to make as few changes as possible.

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java (hunk 13 above)
Comment (simonati) [reply flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java:375]: Please ignore my comment I misread this change. It is completely fine in the way you done it originally.

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java (hunk 7 above)
Comment (simonati) [reply flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java:502]: You can save the cost of string concatenation with {} if log level doesn't match (which is already a gain). In java before calling any function, its argument list is evaluated. So JVM will execute consumerAndRecords.get().getOffsetMapString() statement before invoking logger.trace(arg1, arg2) to make sure arg2 is computed so that it can be passed into.

Comment (simonati) [reply flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java:502]: You can save the cost of string concatenation with {} if log level doesn't match (which is already a gain). In java before calling any function, its argument list is evaluated. So JVM will execute consumerAndRecords.get().getOffsetMapString() statement before invoking logger.trace(arg1, arg2) to make sure arg2 is computed so that it can be passed into.

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java (hunk 13 above)
Comment (simonati) [reply flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java:375]: Please ignore my prev comment I misread this change. Sorry for that. It is completely fine in the way you done it originally.

File: flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java (hunk 8 above)
Comment (simonati) [reply flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java:157]: Indeed. conter.stop() logs this data at info level already.

**Review by bessbd — APPROVED** [review]
LGTM, +1

**Comment by bessbd** [comment]
I'm about to commit this

**Comment by bessbd** [comment]
@jholoman : thank you for the patch!

**Comment by bessbd** [comment]
@simonati : thank you for the reviews!

**Comment by JerryShaoPro** [comment]
I also found this bug when an agent down and after kafka rebalance. After read the code of the commit offsets part, I found offsets map has not been clear, when the cousumer's partitions changed. I'd like to commit a patch to fix this bug, but found it has been fixed by others, haha.

Thanks all of the above persons.

**Comment by qqj1979** [comment]
- KafkaSource EXCEPTION, {}
org.apache.kafka.clients.consumer.CommitFailedException: Commit cannot be completed since the group has already rebalanced and assigned the partitions to another member. This means that the time between subsequent calls to poll() was longer than the configured session.timeout.ms, which typically implies that the poll loop is spending too much time message processing. You can address this either by increasing the session timeout or by reducing the maximum size of batches returned in poll() with max.poll.records.

**I encountered this problem, how to solve the solution。**
